### PR TITLE
rgw: http client cleanup, streaming writes

### PR DESCRIPTION
**Pull request by yehudasa**
- remove the non-threaded case for http manager
 - have a global http manager instance (used instead of the non-threaded case)
 - api modifications, cleanups
 - new http client api that supports streaming writes

### CODE REVIEW

**Comment by ZVampirEM77** [comment]
@yehudasa at first, whether we should merge this PR into RGW before cloud sync or not? I think streaming writes is an independent function and it is useful not only for cloud sync.

**Comment by liewegas** [comment]
@yehudasa ping

**Comment by yehudasa** [comment]
@ZVampirEM77 it is important, but the only user for it currently is the cloud sync work, and things are still in flux, so I'm not sure I want to merge it in as is. OTOH,

File: src/rgw/rgw_http_client.cc [hunk 1]
```diff
@@ -380,7 +435,7 @@ int RGWHTTPClient::wait()
 RGWHTTPClient::~RGWHTTPClient()
 {
   if (req_data) {
-    RGWHTTPManager *http_manager = req_data->get_manager();
+    RGWHTTPManager *http_manager = req_data->mgr;
```
Comment (ZVampirEM77) [inline src/rgw/rgw_http_client.cc]: @yehudasa I see in `rgw_http_req_data::get_manager()` the mutex::locker is holded, why we no more need it now?

**Comment by theanalyst** [comment]
@yehudasa isn't this already a part of https://github.com/ceph/ceph/pull/18932

**Pull request closed without merging**